### PR TITLE
do not let user override hadoop job settings explicitly provided by druid code

### PR DESCRIPTION
**Pull request by himanshug**

### CODE REVIEW

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 1]
```diff
@@ -139,6 +140,18 @@ public OutputStream getOutput() throws IOException
     }
   }
 
+  public static void addJobProperties(Configuration conf, Map<String, String> properties)
+  {
+    for (final Map.Entry<String, String> entry : properties.entrySet()) {
+      String prop = entry.getKey();
+      if (prop.equals(JobContext.NUM_REDUCES)) {
+        log.warn("Map-Reduce propery [%s] is ignored.", prop);
```
Comment (fjy) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: property*

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 2]
```diff
@@ -148,7 +161,12 @@ public static Configuration injectSystemProperties(Configuration conf)
   {
     for (String propName : System.getProperties().stringPropertyNames()) {
       if (propName.startsWith("hadoop.")) {
-        conf.set(propName.substring("hadoop.".length()), System.getProperty(propName));
+        String prop = propName.substring("hadoop.".length());
+        if (prop.equals(JobContext.NUM_REDUCES)) {
+          log.warn("Map-Reduce propery [%s] is ignored.", prop);
```
Comment (fjy) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: same error here

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 3]
```diff
@@ -148,7 +161,12 @@ public static Configuration injectSystemProperties(Configuration conf)
   {
     for (String propName : System.getProperties().stringPropertyNames()) {
       if (propName.startsWith("hadoop.")) {
-        conf.set(propName.substring("hadoop.".length()), System.getProperty(propName));
+        String prop = propName.substring("hadoop.".length());
```
Comment (gianm) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Can this logic be combined with the other method doing a similar thing, maybe like a "maybeAddJobProperty" method that both things can call?

**Comment by gianm** [comment]
@himanshug, is this fixing a bug or just general hygiene?

Asking because I have one concern: IIRC most of our hadoop jobs explicitly set a number of reducers (so usually mapreduce.job.reduces doesn't do anything anyway) except for the group-by job under single dimension partitioning. That job needs some way for users to specify the number of reducers, or else it will default to 1 reducer and performance will be poor. mapreduce.job.reduces seems like as good a way as any, especially since I don't think we can necessarily pick a good default for that job.

**Comment by himanshug** [comment]
@gianm I thought in all the cases druid wanted to setup number of reducers itself (e.g. IndexGeneratorJob sets the number of reducers but will get overridden by user setting and can potentially cause indexing to faile e.g. if the users sets the number to 0).
If user overriding num reducers is a genuine usecase then this PR is unnecessary and can be closed.

**Comment by gianm** [comment]
@himanshug in that case how do you feel about moving the "addJobProperties" step to right after "injectSystemProperties" for all the hadoop jobs? Currently it's a lot later than that, which does seem like a bug, since it means you can overwrite things that are explicitly set in the code (and we probably don't want to allow that).

The groupByJob in DeterminePartitionsJob does not explicitly set a number of reducers, so it does need some way to set it. I think for most people it's going to use whatever they have in their mapred-site.xml.

**Comment by himanshug** [comment]
@gianm sounds like in some cases we do want user to be able to specify num reducers so I will just update IndexGeneratorJob to not allow overrides.

**Comment by himanshug** [comment]
@gianm @fjy changed PR to let user properties added first and then by druid code so that stuff is picked up from user but overridden by druid (if necessary)

**Comment by gianm** [comment]
:+1:

**Pull request merged**